include/libp2p/basic/scheduler.hpp

```diff
@@ -0,0 +1,239 @@
+     * If abs_time == 0 then SchedulerBackendFeedback::pulse()
+     * will be called on the next IO loop cycle with zero argument
+     * @param scheduler Weak reference to scheduler (as it may expire)
+     */
+    virtual void setTimer(
+        std::chrono::milliseconds abs_time,
+        std::weak_ptr<SchedulerBackendFeedback> scheduler) = 0;
+  };
+
+  /**
+   * Scheduler API. Provides callback deferring facilities and low-res timer
+   */
+  class Scheduler : public SchedulerBackendFeedback,
+                    public std::enable_shared_from_this<Scheduler> {
+   public:
+    struct Config {
+      static constexpr std::chrono::milliseconds kMaxTimerDelay =
+          std::chrono::milliseconds(10);
+
+      /**
+       * This delay is needed to avoid too often timer switches.
+       * Performance concerns
+       */
+      std::chrono::milliseconds max_timer_delay = kMaxTimerDelay;
+    };
+
+    enum class Error {
+      // Invalid argument passed
+      kInvalidArgument = 1,
+
+      // Scheduler handle detached, cannot reschedule
+      kHandleDetached,
+
+      // Scheduler item not found, cannot reschedule
+      kItemNotFound,
+    };
+
+    /**
+     * Handle provides scoped callbacks lifetime, allows for canceling and
+     * rescheduling
+     */
+    class Handle {
+     public:
+      /**
+       * { async-time, seq-number } structure for proper ordering and uniqueness
+       * within Scheduler
+       */
+      using Ticket = std::pair<int64_t, uint64_t>;
+
+      Handle(const Handle &) = delete;
+      Handle &operator=(const Handle &) = delete;
+
+      Handle() = default;
+      Handle(Handle &&) = default;
+
+      /**
+       * Ctor. Called from SchedulerImpl
+       * @param ticket unique ticket, used for cancelling and rescheduling
+       * @param scheduler handle's owner object
+       */
+      Handle(Ticket ticket, std::weak_ptr<Scheduler> scheduler);
+
+      /**
+       *  Move assignment cancels existing ticket, if any
+       */
+      Handle &operator=(Handle &&r) noexcept;
+
+      /**
+       *  Non-trivial dtor cancels existing ticket, if any
+       */
+      ~Handle();
+
+      /**
+       * Cancels existing ticket, if any
+       */
+      void cancel() noexcept;
+
+      /**
+       * Reschedules existing ticket, if it is still active.
+       * Allows reentrancy (can be called from inside the callback)
+       * @param delay_from_now Relative time
+       * @return success or error value from Scheduler::Error
+       */
+      outcome::result<void> reschedule(
+          std::chrono::milliseconds delay_from_now) noexcept;
+
+     private:
+      Ticket ticket_;
+      std::weak_ptr<Scheduler> scheduler_;
+    };
+
+    using Callback = std::function<void()>;
+
+    /**
+     * Defers callback to be executed during the next IO loop cycle
+     * @param cb callback
+     */
+    void schedule(Callback &&cb) noexcept {
+      std::ignore = scheduleImpl(std::forward<Callback>(cb),
```

Comment:
```suggestion
      std::ignore = scheduleImpl(std::move(cb),
```
`Callback &&cb` is not a [forwarding reference](https://en.cppreference.com/w/cpp/language/reference#Forwarding_references).